src/persist/src/gen.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Generated protobuf code and companion impls.
+
+#![allow(missing_docs)]
+
+include!(concat!(env!("OUT_DIR"), "/protobuf/mod.rs"));
+
+use std::io::Read;
+
+use md5::{Digest, Md5};
+use ore::cast::CastFrom;
+use protobuf::Message;
+
+use crate::error::Error;
+use crate::gen::persist::ProtoMeta;
+
+impl ProtoMeta {
+    /// A versioning for how we store the protobuf serialization.
+    ///
+    /// Protobuf handles most of our backward and forward compatibility, but we
+    /// don't just store the raw protobuf message serialization. This version
+    /// number determines the exactly what that format is.
+    ///
+    /// Once we commit to backward compatibility, this should only change if we
+    /// decide to e.g. switch from protobuf to some other encoding entirely.
+    ///
+    /// All versions less than this were developmental. If encountered, it's
+    /// safe to delete all data in blob storage. If a greater version is seen,
+    /// then some major change has happened and this code has no idea what is
+    /// going on and should refuse to touch it.
+    ///
+    /// The following is an EBNF-ish spec for the format:
+    ///
+    /// ```none
+    /// encoding = 6u8 v6_encoding
+    /// v6_encoding = md5_checksum proto_meta
+    /// md5_checksum = u8 u8 u8 u8 (little endian, md5 of proto_meta)
+    /// proto_meta = u8* (the protobuf serialization of ProtoMeta)
+    /// ```
+    pub const ENCODING_VERSION: u8 = 7;
+
+    /// The [Self::ENCODING_VERSION] of this previously encoded ProtoMeta.
+    ///
+    /// Returns an error if the input is malformed.
+    pub fn encoded_version(buf: &[u8]) -> Result<u8, Error> {
```

Comment:
just to confirm: we don't want to keep the md5 hash as a proto field because from skimming the encoding guide it seems like theres no stable order to how fields are serialized, so its a bit chicken and egg -- you've got the encoded struct + you computed the md5 (with the md5 field blanked out to 16 0 bytes) but now theres no way to know where to place the md5 in the encoded data -- is that right?

but it seems like the encoding version could be a field in the proto spec? is that right or is there a more fundamental reason we don't want to do that? I guess for now we need a way to grab the encoding from the first byte, but that seems more like a short term hack and something we could work around

Reply:
> just to confirm: we don't want to keep the md5 hash as a proto field because from skimming the encoding guide it seems like theres no stable order to how fields are serialized, so its a bit chicken and egg -- you've got the encoded struct + you computed the md5 (with the md5 field blanked out to 16 0 bytes) but now theres no way to know where to place the md5 in the encoded data -- is that right?

oh that's clever. it's easy enough to hack this, the proto encoding says that if you duplicate a non-repeated field id, the last one wins. (said another way, you can "merge" two encoded protos by simply concatenating them) so we could serialize this, checksum it, and then manually encode that data and append it to the end. i really want to get this in before the next release is cut, so mind if we leave this as a TODO (might actually be a good one for you to pick off to get more familiar with protobuf)

Reply:
> but it seems like the encoding version could be a field in the proto spec? is that right or is there a more fundamental reason we don't want to do that? I guess for now we need a way to grab the encoding from the first byte, but that seems more like a short term hack and something we could work around

ah damn, i tried to preemptively address this exact Q with comments but looks like I failed :). the encoding version stays outside because it's an escape hatch in case way down the line we want to stop using proto for this and use something else entirely (_extremely_ unlikely but we'll be glad to have it if it comes up)

Reply:
ive heard this md5 trick before, can't claim to have come up with that myself :D - no worries on doing this right away totally fine to leave it as a todo

Reply:
so, re: encoding version, just to say it back to you -- it sounds like we want _some_ identifier of what kind of data this is in a stable location, ideally near the front of the serialized data, so that down the line, we could use this identifier + location to determine what kind of data we're looking at, and the proto spec doesn't give us a way to express "I'd like encoded bytes [start..end] to be foo please" and so we have to handroll?

If so, makes sense and I guess its the best we can do. I'll defer to your judgement here because. The small downside i noticed is that it seems like we're currently encoding into a vec, and then copying its contents into another buffer, which is a small amount of extra copying, but im sure theres ways to work around that,

Reply:
yeah that's right. good point on the temp vec, even without pushing the checksum into proto, that's not necessary if we put the checksum at the end, so i did that. i tried to hack in actually computing the checksum as we go and it was _just_ hairy enough that I left that part as a TODO, but it won't be a breaking change